### PR TITLE
ovl_En_Test2 Decompiled

### PR DESCRIPTION
**Pull request by sonicdcer**

### CODE REVIEW

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 1]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+    OBJECT_DEKUCITY_ANA_OBJ, OBJECT_SICHITAI_OBJ,  OBJECT_YUKIMURA_OBJ, OBJECT_HAKUGIN_OBJ, OBJECT_HAKUGIN_OBJ,
+    OBJECT_HAKUGIN_OBJ,      OBJECT_MEGANEANA_OBJ, OBJECT_HAKA_OBJ,     OBJECT_HAKA_OBJ,    OBJECT_HAKUGIN_OBJ,
+    OBJECT_HAKUGIN_OBJ,      OBJECT_HAKUGIN_OBJ,   OBJECT_HAKUGIN_OBJ,
+};
+
+void EnTest2_Init(Actor* thisx, PlayState* play) {
+    EnTest2* this = THIS;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    if ((this->actor.params == 0xB) || (this->actor.params == 0xC)) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:75]: These need an enum.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 2]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+    }
+}
+
+void EnTest2_Update(Actor* thisx, PlayState* play) {
+    s32 pad;
+    s32 objectIndex;
+    EnTest2DisplayObjs* dispObjects;
+    EnTest2* this = THIS;
+    objectIndex = Object_GetIndex(&play->objectCtx, sTest2Objects[this->actor.params]);
+
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:86]:
```suggestion

    objectIndex = Object_GetIndex(&play->objectCtx, sTest2Objects[this->actor.params]);
```
Space between declarations and code.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 3]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+    if (objectIndex < 0) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+    if (Object_IsLoaded(&play->objectCtx, objectIndex)) {
+        dispObjects = &sDisplayObjects[this->actor.params];
+        this->actor.objBankIndex = objectIndex;
+        this->actor.draw = EnTest2_Draw;
+        if ((dispObjects->animMat) != 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:95]:
```suggestion
        if (dispObjects->animMat != NULL) {
```
Pointer, compare to NULL.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 4]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+            Actor_SetObjectDependency(play, &this->actor);
+            this->animMat = Lib_SegmentedToVirtual(dispObjects->animMat);
+        }
+        if (play->roomCtx.currRoom.unk5 != 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:99]: IIRC this is boolean... @engineer124 will remember.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 5]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+        } else {
+            this->actor.update = Actor_Noop;
+        }
+    }
+}
+
+void EnTest2_SetActorFlags(Actor* thisx, PlayState* play) {
+    EnTest2* this = THIS;
+
+    if (play->actorCtx.unk4 == 0x64) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:110]:
```suggestion
    if (play->actorCtx.unk4 == 100) {
```
although the magic PR will change this anyway.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 6]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+    }
+    if CHECK_FLAG_ALL (this->actor.flags, ACTOR_FLAG_80) {
+        OPEN_DISPS(play->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Test2/EnTest2_Update.s")
+        func_8012C2DC(play->state.gfxCtx);
+        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(play->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Test2/func_80A3E4EC.s")
+        if (dispObjects != 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:131]: Compare to NULL

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 7]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+            gSPDisplayList(POLY_XLU_DISP++, dispObjects);
+        }
+        gSPDisplayList(POLY_XLU_DISP++, sDisplayObjects[this->actor.params].DList1);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Test2/func_80A3E524.s")
+        CLOSE_DISPS(play->state.gfxCtx);
+    } else {
+        if (dispObjects != 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:138]: NULL

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 8]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+
+static InitChainEntry sInitChain[] = {
     ICHAIN_VEC3F(scale, 1, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneForward, 8000, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneScale, 800, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneDownward, 2500, ICHAIN_STOP),
 };
 
-#endif
+static s16 sTest2Objects[] = {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:65]:
```suggestion
static s16 sObjectIds[] = {
```
is usual

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 9]
```diff
@@ -5,15 +5,28 @@
...
 #define THIS ((EnTest2*)thisx)
 
 void EnTest2_Init(Actor* thisx, PlayState* play);
 void EnTest2_Update(Actor* thisx, PlayState* play);
+void EnTest2_SetActorFlags(Actor* thisx, PlayState* play);
+void EnTest2_Draw(Actor* thisx, PlayState* play);
+
+typedef struct EnTest2DisplayObjs {
+    Gfx* DList1;
+    Gfx* DList2;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:26]: Should be camelCase. Also don't need to name the struct at the beginning for something that's only used once like this.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 10]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
     (ActorFunc)NULL,
 };
 
-// static InitChainEntry sInitChain[] = {
-static InitChainEntry D_80A3E74C[] = {
+EnTest2DisplayObjs sDisplayObjects[] = {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:42]: I'd just call it `sModelInfo`, which we've used before.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 11]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+void EnTest2_Draw(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnTest2* this = THIS;
+    Gfx* dispObjects = sDisplayObjects[this->actor.params].DList2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Test2/EnTest2_Init.s")
+    if (this->animMat != NULL) {
+        AnimatedMat_Draw(play, this->animMat);
+    }
+    if CHECK_FLAG_ALL (this->actor.flags, ACTOR_FLAG_80) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:125]: Something's gone wrong with this if.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 12]
```diff
@@ -26,22 +39,105 @@ const ActorInit En_Test2_InitVars = {
...
+    } else {
+        this->actor.flags &= ~ACTOR_FLAG_80;
+    }
+}
 
-extern InitChainEntry D_80A3E74C[];
+void EnTest2_Draw(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnTest2* this = THIS;
+    Gfx* dispObjects = sDisplayObjects[this->actor.params].DList2;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:120]: This is a single displaylist, `dList` will do if you don't want to check what it actually is.

File: assets/xml/objects/object_hakugin_obj.xml [hunk 13]
```diff
@@ -6,37 +6,41 @@
...
         <DList Name="object_hakugin_obj_DL_0016D0" Offset="0x16D0" />
         <DList Name="object_hakugin_obj_DL_0016D8" Offset="0x16D8" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_0017A0" Size="0x10" Offset="0x17A0" /> -->
+        <TextureAnimation Name="object_hakugin_obj_Matanimheader_0017A8" Offset="0x17A8" />
         <Texture Name="object_hakugin_obj_TLUT_0017B0" OutName="tlut_0017B0" Format="rgba16" Width="4" Height="4" Offset="0x17B0" />
         <Texture Name="object_hakugin_obj_Tex_0017D0" OutName="tex_0017D0" Format="ci4" Width="64" Height="64" Offset="0x17D0" />
         <DList Name="object_hakugin_obj_DL_002010" Offset="0x2010" />
         <DList Name="object_hakugin_obj_DL_002018" Offset="0x2018" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_0020D8" Size="0x18" Offset="0x20D8" /> -->
+        <!-- <Blob Name="object_hakugin_obj_Blob_0020D8" Size="0x8" Offset="0x20D8" /> -->
```
Comment (EllipticEllipsis) [inline assets/xml/objects/object_hakugin_obj.xml:19]: This is just zeros and can be removed.

File: assets/xml/objects/object_hakugin_obj.xml [hunk 14]
```diff
@@ -6,37 +6,41 @@
...
+        <TextureAnimation Name="object_hakugin_obj_Matanimheader_0020E8" Offset="0x20E8" />
         <Texture Name="object_hakugin_obj_TLUT_0020F0" OutName="tlut_0020F0" Format="rgba16" Width="4" Height="4" Offset="0x20F0" />
         <Texture Name="object_hakugin_obj_Tex_002110" OutName="tex_002110" Format="ci4" Width="64" Height="64" Offset="0x2110" />
         <DList Name="object_hakugin_obj_DL_002970" Offset="0x2970" />
         <DList Name="object_hakugin_obj_DL_002978" Offset="0x2978" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_002A48" Size="0x18" Offset="0x2A48" /> -->
+        <!-- <Blob Name="object_hakugin_obj_Blob_002A48" Size="0x10" Offset="0x2A48" /> -->
```
Comment (EllipticEllipsis) [inline assets/xml/objects/object_hakugin_obj.xml:25]: As is this (and the size looks wrong anyway).

File: assets/xml/objects/object_hakugin_obj.xml [hunk 15]
```diff
@@ -6,37 +6,41 @@
...
+        <TextureAnimation Name="object_hakugin_obj_Matanimheader_002A58" Offset="0x2A58" />
         <DList Name="object_hakugin_obj_DL_0035A0" Offset="0x35A0" />
         <DList Name="object_hakugin_obj_DL_0035A8" Offset="0x35A8" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_003878" Size="0x18" Offset="0x3878" /> -->
+        <!-- <Blob Name="object_hakugin_obj_Blob_003878" Size="0x8" Offset="0x3878" /> -->
```
Comment (EllipticEllipsis) [inline assets/xml/objects/object_hakugin_obj.xml:29]: Same here, just zero padding, can be removed.

File: assets/xml/objects/object_hakugin_obj.xml [hunk 16]
```diff
@@ -6,37 +6,41 @@
...
         <DList Name="object_hakugin_obj_DL_004920" Offset="0x4920" />
         <DList Name="object_hakugin_obj_DL_004928" Offset="0x4928" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_0049F0" Size="0x10" Offset="0x49F0" /> -->
+        <TextureAnimation Name="object_hakugin_obj_Matanimheader_0049F8" Offset="0x49F8" />
         <Texture Name="object_hakugin_obj_TLUT_004A00" OutName="tlut_004A00" Format="rgba16" Width="4" Height="4" Offset="0x4A00" />
         <Texture Name="object_hakugin_obj_Tex_004A20" OutName="tex_004A20" Format="ci4" Width="64" Height="64" Offset="0x4A20" />
         <DList Name="object_hakugin_obj_DL_005260" Offset="0x5260" />
         <DList Name="object_hakugin_obj_DL_005268" Offset="0x5268" />
-        <!-- <Blob Name="object_hakugin_obj_Blob_005328" Size="0x18" Offset="0x5328" /> -->
+        <!-- <Blob Name="object_hakugin_obj_Blob_005328" Size="0x8" Offset="0x5328" /> -->
```
Comment (EllipticEllipsis) [inline assets/xml/objects/object_hakugin_obj.xml:42]: And here.

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c (hunk 11 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Test2/z_en_test2.c:125]: Ifs in C are required to have ().

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 17]
```diff
@@ -5,15 +5,33 @@
...
 
 void EnTest2_Init(Actor* thisx, PlayState* play);
 void EnTest2_Update(Actor* thisx, PlayState* play);
+void EnTest2_SetActorFlags(Actor* thisx, PlayState* play);
+void EnTest2_Draw(Actor* thisx, PlayState* play);
+
+typedef enum {
+    /* 0xB */ EN_TEST2_PARAMS_B = 0xB,
+    /* 0xC */ EN_TEST2_PARAMS_C
+} EnTest2_Params;
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:27]: .h file please :)

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 18]
```diff
@@ -5,15 +5,33 @@
...
+
+typedef struct EnTest2DisplayObjs {
+    Gfx* dList1;
+    Gfx* dList2;
+    AnimatedMaterial* animMat;
+} EnTest2DisplayObjs;
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:33]:
```suggestion
typedef struct EnTest2ModelInfo {
    /* 0x0 */ Gfx* dList1;
    /* 0x4 */ Gfx* dList2;
    /* 0x8 */ AnimatedMaterial* animMat;
} EnTest2ModelInfo; // size = 0xC
```

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 19]
```diff
@@ -26,22 +44,105 @@ const ActorInit En_Test2_InitVars = {
...
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    if ((this->actor.params == EN_TEST2_PARAMS_B) || (this->actor.params == EN_TEST2_PARAMS_C)) {
+        this->actor.flags |= ACTOR_FLAG_20;
+    }
+}
+
+void EnTest2_Update(Actor* thisx, PlayState* play) {
+    s32 pad;
+    s32 objectIndex;
+    EnTest2DisplayObjs* dispObjects;
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:88]:
```suggestion
    EnTest2DisplayObjs* modelInfo;
```

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 20]
```diff
@@ -26,22 +44,105 @@ const ActorInit En_Test2_InitVars = {
...
+    objectIndex = Object_GetIndex(&play->objectCtx, sObjectIds[this->actor.params]);
+    if (objectIndex < 0) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+    if (Object_IsLoaded(&play->objectCtx, objectIndex)) {
+        dispObjects = &sModelInfo[this->actor.params];
+        this->actor.objBankIndex = objectIndex;
+        this->actor.draw = EnTest2_Draw;
+        if ((dispObjects->animMat) != NULL) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:100]:
```suggestion
        if (dispObjects->animMat != NULL) {
```

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 21]
```diff
@@ -26,22 +44,105 @@ const ActorInit En_Test2_InitVars = {
...
+void EnTest2_Draw(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnTest2* this = THIS;
+    Gfx* dList = sModelInfo[this->actor.params].dList2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Test2/EnTest2_Init.s")
+    if (this->animMat != NULL) {
+        AnimatedMat_Draw(play, this->animMat);
+    }
+    if CHECK_FLAG_ALL (this->actor.flags, ACTOR_FLAG_80) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:130]: As mentioned by Elliptic above, the `if` itself needs `()`

File: src/overlays/actors/ovl_En_Test2/z_en_test2.c [hunk 22]
```diff
@@ -5,15 +5,28 @@
...
 void EnTest2_Init(Actor* thisx, PlayState* play);
 void EnTest2_Update(Actor* thisx, PlayState* play);
+void EnTest2_SetActorFlags(Actor* thisx, PlayState* play);
+void EnTest2_Draw(Actor* thisx, PlayState* play);
+
+typedef struct EnTest2ModelInfo {
+    /* 0x0 */ Gfx* dList1;
+    /* 0x4 */ Gfx* dList2;
+    /* 0x8 */ AnimatedMaterial* animMat;
+} EnTest2ModelInfo;
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Test2/z_en_test2.c:28]:
```suggestion
} EnTest2ModelInfo; // size = 0xC
```

**Review by hensldm — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**